### DISCUSSION TITLE
Csla6 BusinessBindingListBase EFCore Save()

### DISCUSSION THREAD

**Discussion by Brad-IMS** [MarimerLLC/csla#3276] [Questions] [answered]
We have a Csla business library with about fifty classes, mostly based on SwitchableObject, that we have been using for some years.  The technology is Csla 4.51, .Net Framework 4.8, and ADO.Net.  We are in the process of creating a Blazor Server app, and need to upgrade to Csla 6.2.2, .Net6, and EFCore, using Dependency Injection.  From various sources I have pieced together a fully-working switchable object class based on BusinessBase, and a partly functional collection class based on BusinessBindingListBase.  The CreateChild and FetchChild appear to be working correctly, but I have been unable to find a sample of UpdateChild to crib from.  When I execute this code:

	POSList posList = posListPortal.FetchChild();
	posList[0].VendorID = "Sears";
	posList = posList.Save();
	
Everything works correctly until the 'Save'.  UpdateChild in my collection class, POSList, is never touched, and instead I receive an error message "Can not directly save a child object".  However, it's the collection I'm trying to save, not an individual object.

This is my collection class:

    //##################################################################################################
    // Bindable collection of POS objects
    // Based on Rocky Lhotka's open source CSLA6 Framework, Copyright Marimer LLC
    //--------------------------------------------------------------------------------------------------
    [Serializable]
    public class POSList : BusinessBindingListBase<POSList, POS>
    {
        #region Properties

        public int ReturnCode { get; set; } = -99;
        public string? ReturnMessage { get; set; } = "Initialized.";

        #endregion  // Properties

        #region Data Access

        [CreateChild]
        private void Create([Inject] DataAccess.IPOSDal dal, [Inject] IChildDataPortal<POS> posListPortal)
        {
            try
            {
                Add(posListPortal.CreateChild());
                ReturnCode = 0;
                ReturnMessage = "Operation successful.";
            }
            catch (Exception ex)
            {
                ReturnCode = -2;
                ReturnMessage = "Error - " + ex.Message;
            }
        }  // CreateChild

        [FetchChild]
        private void Fetch([Inject] DataAccess.IPOSDal dal, [Inject] IChildDataPortal<POS> posListPortal)
        {
            try
            {
                using (LoadListMode)
                {
                    var data = dal.Get().Select(d => posListPortal.FetchChild(d));
                    AddRange(data);
                    ReturnCode = 0;
                    ReturnMessage = "Operation successful.";
                }
            }
            catch (Exception ex)
            {
                ReturnCode = -2;
                ReturnMessage = "Error - " + ex.Message;
            }
        }  // FetchChild

        [UpdateChild]
        [Transactional(TransactionalTypes.TransactionScope)]
        protected void Child_Update([Inject] DataAccess.IPOSDal dal, [Inject] IChildDataPortal<POS> posListPortal)
        {
            try
            {
                base.Child_Update(this);
                ReturnCode = 0;
                ReturnMessage = "Operation successful.";
            }
            catch (Exception ex)
            {
                ReturnCode = -2;
                ReturnMessage = "Error - " + ex.Message;
            }
        }  // UpdateChild

        #endregion  // Data Access

    }

And these are the child functions in the single-object POS class:

	#region Child Data Access

	[RunLocal]
	[CreateChild]
	private void CreateChild()
	{
		using (BypassPropertyChecks)
		{
			POSID = " ";
			VendorID = " ";
			LocationID = " ";
			TranTypeDefault = " ";
			ReturnCode = -99;
			ReturnMessage = "Initialized";
		}
		base.Child_Create();
		BusinessRules.CheckRules();
	}  // CreateChild

	[FetchChild]
	private void Fetch(POSEntity childData, [Inject] IPOSDal dal)
	{
		using (BypassPropertyChecks)
		{
			this.POSID = childData.POSID;
			this.VendorID = childData.VendorID;
			this.LocationID = childData.LocationID;
			this.TranTypeDefault = childData.TranTypeDefault;
			this.ReturnCode = 0;
			this.ReturnMessage = childData.ReturnMessage;
		}
	}  // FetchChild

	[InsertChild]
	private void InsertChild(object parent, [Inject] IPOSDal dal)
	{
		using (BypassPropertyChecks)
		{
			var data = new DataAccess.POSEntity();
			data.POSID = this.POSID;
			data.VendorID = this.VendorID;
			data.LocationID = this.LocationID;
			data.TranTypeDefault = this.TranTypeDefault;
			var result = dal.Insert(data);
			POSID = result.POSID;
		}
	}  // InsertChild

	[UpdateChild]
	private void UpdateChild(object parent, [Inject] IPOSDal dal)
	{
		using (BypassPropertyChecks)
		{
			var data = new DataAccess.POSEntity();
			data.POSID = this.POSID;
			data.VendorID = this.VendorID;
			data.LocationID = this.LocationID;
			data.TranTypeDefault = this.TranTypeDefault;
			dal.Update(data);
		}
	}  // UpdateChild

	[DeleteSelf]
	private void DeleteSelfChild(object parent, [Inject] IPOSDal dal)
	{
		Delete(ReadProperty(POSIDProperty), dal);
	}  // DeleteSelf

	#endregion  // Child Data Access
	
Except for the added 'Child' on the attributes and the added 'object Parent' in the signatures, these are identical to the non-child functions that are working correctly.  So, my questions are:

1. What is the correct code for the POSList UpdateChild function?
2. Are there any missing functions in my POSList class?
3. Is Save() still the correct way to invoke updates on a collection class?
4. Am I missing a DeleteChild function in my POS class?

**Comment by rockfordlhotka** [comment, marked as answer]
What is the type of your `IDataPortal<T>`? It kind of sounds like you are using an `IDataPortal<Child>` instead of `IDataPortal<ChildList>`?

*Update:* actually it looks like you create the list _as a child_ by calling `FetchChild`.

At some point you _must_ use the root data portal to get the top-level root object, and that is the option that you need to save. You can't directly save an object that you created via `FetchChild`.

**Reply by rockfordlhotka** [reply]
I think this is the basis of what you are after?

```c#
using Csla;
using Csla.Configuration;
using Microsoft.Extensions.DependencyInjection;
using System.Diagnostics;

var services = new ServiceCollection();
services.AddCsla();
var provider = services.BuildServiceProvider();

var parentDp = provider.GetRequiredService<IDataPortal<Parent>>();
var parent = parentDp.Fetch();
Debug.Assert(parent != null);
Debug.Assert(parent.ChildList != null);
Debug.Assert(parent.ChildList.IsChild);
await parent.SaveAndMergeAsync();

var childListDp = provider.GetRequiredService<IDataPortal<ChildList>>();
var childList = childListDp.Fetch();
Debug.Assert(childList != null);
Debug.Assert(!childList.IsChild);
await childList.SaveAndMergeAsync();

Console.WriteLine("Completed");
Console.ReadLine();


[Serializable]
public class Parent : BusinessBase<Parent>
{
  public static readonly PropertyInfo<ChildList> ChildListProperty = RegisterProperty<ChildList>(nameof(ChildList));
  public ChildList ChildList
  {
    get => GetProperty(ChildListProperty);
    set => SetProperty(ChildListProperty, value);
  }

  [Fetch]
  private void Fetch([Inject] IChildDataPortal<ChildList> dp)
  {
    ChildList = dp.FetchChild();
  }

  [Insert][Update]
  private void InsertUpdate([Inject] IChildDataPortal<ChildList> dp)
  {
    dp.UpdateChild(ChildList);
  }
}

[Serializable]
public class ChildList : BusinessListBase<ChildList, Child>
{
  [Fetch]
  private void Fetch()
  { }

  [FetchChild]
  private void FetchChild() { }

  [Update]
  private void Update()
  {
    base.Child_Update();
  }

  [UpdateChild]
  private void UpdateChild()
  {
    base.Child_Update();
  }
}

[Serializable]
public class Child : BusinessBase<Child>
{
  [FetchChild]
  private void FetchChild() { }

  [InsertChild][UpdateChild] private void InsertChild() { }
}
```

**Reply by Brad-IMS** [reply]
I suspect your first comment is the root of my problem.  I'll need to dig into it,  But your example still has me confused.  I don't see any Delete functions, and it looks as if you have the List as the switchable object rather than the Child.  In our existing system that Parent object does not exist in most cases.  The collection is the parent and the Child object can be created either as a child or a root object.

**Reply by danielmartind** [reply]
I think using your example, if you have the collection as parent your dataportal attributes need to be Create, Fetch and Update and your switchable child would need to add Create, Fetch, Update and Delete. There is a DeleteChildSelf too.

**Reply by Brad-IMS** [reply]
Thanks for the info on the collection attributes.  My child class does have the functions you list, as well as matching child functions.  I'm assuming this hasn't changed from Version 4, where the child functions had a slightly different signature because they were passed the parent object.  For example:

        [Update]
        private void Update([Inject] IPOSDal dal)

        [UpdateChild]
        private void UpdateChild(object parent, [Inject] IPOSDal dal)

If the parent object is no longer being passed, I could use the same function for both root and child operations.   Is this the case?

Also, I realized afterward I should have included all of the calling class, to show how it was set up:

```
   class Program
    {
        static async Task Main(string[] args)
        {
            var services = new ServiceCollection();
            services.AddCsla();
            services.AddTransient<Program>();
            services.AddDbContext<_MyDbContext>(options => options.UseSqlServer("<connectionstring>"));
            services.AddTransient(typeof(DataAccess.IPOSDal), typeof(DataAccess.POSDal));
            var provider = services.BuildServiceProvider();
            var program = provider.GetRequiredService<Program>();
            await program.Run();
        }

        // These are all defined because I'm not sure which one I should be using
        private IDataPortal<POS> posPortal;
        private IDataPortal<POSList> posListPortal;
        private IChildDataPortal<POS> posChildPortal;
        private IChildDataPortal<POSList> posListChildPortal;

        public Program(IDataPortal<POS> portal, IDataPortal<POSList> listPortal, IChildDataPortal<POS> childPortal, IChildDataPortal<POSList> listChildPortal)
        {
            posPortal = portal;
            posListPortal = listPortal;
            posChildPortal = childPortal;
            posListChildPortal = listChildPortal;
        }

        public async Task Run()
        {
            //ToDo: Test collection (POSList) update
            POSList posList = posListChildPortal.FetchChild();
            posList[0].VendorID = "Sears";
            posList = posList.Save();
            //Result: Populates collection. Child records are marked as child. Save throws 'Cannot directly update child' message.
        }

    }
```

**Reply by Brad-IMS** [reply]
Problem solved!  I implemented your advice, changing the POSList attributes to Create, Fetch, and Update, and changed which dataportal I was using.  So, the working combination was:

            POSList posList = posListPortal.Fetch();
            posList[0].VendorID = "Sears";
            posList = posList.Save();

**Reply by Brad-IMS** [reply]
For the edification of anyone else struggling with this combination of CSLA6, EFCore, and Dependency Injection, the following is my data access component:

```
    public class POSDal : IPOSDal, IDisposable
    {
        private readonly _MyDbContext db;

        public POSDal(_MyDbContext context)
        {
            db = context;
        }

        public bool Exists(string id)
        {
            var obj = db.POS?.Where(p => p.POSID == id).FirstOrDefault();
            return !(obj == null);
        }

        public POSEntity Get(string id)
        {
            var obj = db.POS?.Where(p => p.POSID == id).FirstOrDefault();
            if (obj != null)
            {
                return obj;
            }
            else
            {
                throw new KeyNotFoundException($"Id {id}");
            }
        }

        public List<POSEntity> Get()
        {
                return db.POS?.Where(r => true)?.ToList();
        }

        public POSEntity Insert(POSEntity obj)
        {
            if (Exists(obj?.POSID))
                throw new InvalidOperationException($"Key exists {obj.POSID}");
            {
                db.POS?.Add(obj);
                db.SaveChanges();
            }
            return obj;
        }

        public POSEntity Update(POSEntity obj)
        {
            var old = Get(obj.POSID);
            Csla.Data.DataMapper.Map(obj, old);
            db.SaveChanges();
            return old;
        }

        public bool Delete(string id)
        {
            var obj = db.POS?.Where(p => p.POSID == id).FirstOrDefault();
            if (obj != null)
            {
                db.POS?.Remove(obj);
                db.SaveChanges();
                return true;
            }
            else
            {
                return false;
            }
        }

        public void Dispose()
        {
            db.Dispose();
        }

    }
```
